examples/checkpoint_traversal.cc

```diff
@@ -0,0 +1,139 @@
+// Questions? Contact darma@sandia.gov
+//
+// *****************************************************************************
+//@HEADER
+*/
+
+#include <checkpoint/checkpoint.h>
+
+#include <cstdio>
+
+namespace checkpoint { namespace examples {
+
+struct TestObject {
+
+  TestObject() = default;
+
+  struct MakeTag { };
+
+  explicit TestObject(MakeTag) {
+    for (int i = 0; i < 10; i++) {
+      vec1.push_back(i);
+      vec2.push_back(i * 29.34);
+      vec3.push_back("hello: " + std::to_string(vec2.back()));
+    }
+  }
+
+  template <typename Serializer>
+  void serialize(Serializer& s) {
+    s | a | b | vec1 | vec2 | vec3;
+  }
+
+private:
+  int a = 29;
+  int b = 36;
+  std::vector<int> vec1;
+  std::vector<double> vec2;
+  std::vector<std::string> vec3;
+};
+
+}} // end namespace checkpoint::examples
+
+/// Custom traverser for printing raw bytes
+struct PrintBytesTraverse : checkpoint::Serializer {
+  PrintBytesTraverse() : checkpoint::Serializer(checkpoint::eSerializationMode::None) { }
+
+  void contiguousBytes(void* ptr, std::size_t size, std::size_t num_elms) {
+    printf("PrintBytesTraverse: size=%lu, num_elms=%lu\n", size, num_elms);
+  }
+};
+
+/// Custom traverser for printing typed ranges
+struct TypedTraverse : checkpoint::Serializer {
+  TypedTraverse() : checkpoint::Serializer(checkpoint::eSerializationMode::None) { }
+
+  template <typename SerializerT, typename T>
+  void contiguousTyped(SerializerT&, T*, std::size_t num_elms) {
+    printf("TypedTraverse: type is %s, num=%lu\n", typeid(T).name(), num_elms);
+  }
+};
+
+/// Custom dispatcher to customizing type-specific traversal behavior
+/// For non-vectors, dispatch normally. For vectors, print the data inside
+template <typename SerializerT, typename T>
+struct CustomDispatch {
+  static void serializeIntrusive(SerializerT& s, T& t) {
+    t.serialize(s);
+  }
+  static void serializeNonIntrusive(SerializerT& s, T& t) {
+    serialize(s, t);
+  }
+};
+
+// std::vector specialization for dispatcher, vector is always non-intrusive so
+// skip that overload
+template <typename SerializerT, typename U>
+struct CustomDispatch<SerializerT, std::vector<U>> {
+  static void serializeNonIntrusive(SerializerT& s, std::vector<U>& t) {
+    printf("Traversing vector: size=%lu\n", t.size());
+    for (std::size_t i = 0; i < t.size(); i++) {
+      printf("\t vector[%lu]=%s", i, std::to_string(t[i]).c_str());
+    }
+  }
+};
+
+
+int main(int, char**) {
+  using namespace checkpoint::examples;
+
+  TestObject my_obj(TestObject::MakeTag{});
+
+  // Traverse my_obj with a custom traverser that prints the bytes
+  checkpoint::dispatch::Traverse::with<TestObject, PrintBytesTraverse>(my_obj);
+
+  // Traverse my_obj with a custom traverser that prints the types and lens
+  checkpoint::dispatch::Traverse::with<TestObject, TypedTraverse>(my_obj);
+
+  // Traverse my_obj with a custom dispatcher
+  checkpoint::dispatch::Traverse::with<TestObject, TypedTraverse, CustomDispatch>(my_obj);
```

Comment:
This is written as a smoke test, without verification of the expected results. Maybe use a traverser that matches up a count of elements in vectors or something?